src/frontend/org/voltdb/calciteadapter/rel/physical/VoltDBAggregate.java

```diff
@@ -0,0 +1,331 @@
+/* This file is part of VoltDB.
+ * Copyright (C) 2008-2018 VoltDB Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.voltdb.calciteadapter.rel.physical;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTrait;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.voltdb.calciteadapter.converter.ExpressionTypeConverter;
+import org.voltdb.calciteadapter.converter.RelConverter;
+import org.voltdb.calciteadapter.converter.RexConverter;
+import org.voltdb.calciteadapter.planner.CalcitePlanningException;
+import org.voltdb.expressions.AbstractExpression;
+import org.voltdb.plannodes.AbstractPlanNode;
+import org.voltdb.plannodes.AggregatePlanNode;
+import org.voltdb.plannodes.HashAggregatePlanNode;
+import org.voltdb.plannodes.NodeSchema;
+import org.voltdb.plannodes.PartialAggregatePlanNode;
+import org.voltdb.types.ExpressionType;
+
+public class VoltDBAggregate extends Aggregate implements VoltDBPhysicalRel {
```

Comment:
I like the idea of L and P shorthand for class names.

I think that serial vs. hash vs. partial are all physical characteristics, since they describe the physical algorithm used to perform the aggregation.  There can be just one logical rel for aggregation.  So I'm proposing these classes:
- VoltDBLAggregate
- VoltDBPSerialAggregate
- VoltDBPHashAggregate
- VoltDBPPartialAggregate (maybe, not sure about this one)
Does this make sense?

Reply:
I understand that this is still WIP though.

Reply:
Sure.
At the moment Hash-To-Logical transformation is happening in the logical space. If we want to go with a single VoltDBLAggregate I need to modify the rule

Reply:
I also applied the same naming convention to the rules VoltDBL*Rule/ VoltDBP*Rule